PWGHF/D2H/Tasks/taskDplus.cxx

```diff
@@ -94,127 +112,217 @@ struct HfTaskDplus {
     registry.add("hPtVsYGen", "MC particles (matched);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
     registry.add("hPtVsYGenPrompt", "MC particles (matched, prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
     registry.add("hPtVsYGenNonPrompt", "MC particles (matched, non-prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    if (doprocessDataWithMl) {
+      registry.add("hSparseMass", "THn for Dplus", HistType::kTHnSparseF, {massbins, ptbins, axisMlScore0, axisMlScore1, axisMlScore2});
+    }
+    if (doprocessMcWithMl) {
+      registry.add("hSparseMassPrompt", "THn for Dplus Prompt", HistType::kTHnSparseF, {massbins, ptbins, axisMlScore0, axisMlScore1, axisMlScore2});
+      registry.add("hSparseMassFD", "THn for Dplus FD", HistType::kTHnSparseF, {massbins, ptbins, axisMlScore0, axisMlScore1, axisMlScore2});
+      registry.add("hSparseMassBkg", "THn for Dplus Bkg", HistType::kTHnSparseF, {massbins, ptbins, axisMlScore0, axisMlScore1, axisMlScore2});
+    }
   }
 
-  void process(soa::Join<aod::HfCand3Prong, aod::HfSelDplusToPiKPi> const& candidates)
+  // Fill histograms of quantities for the reconstructed Dplus candidates
+  /// \param candidate is candidate
+  template <typename T1>
+  void fillHisto(const T1& candidate)
   {
-    for (const auto& candidate : selectedDPlusCandidates) {
-      // not possible in Filter since expressions do not support binary operators
-      if (!(candidate.hfflag() & 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
-        continue;
-      }
+    // not possible in Filter since expressions do not support binary operators
+    if (!(candidate.hfflag() & 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
+      return;
+    }
```

Comment:
It is now possible to do this. You could add a filter taking https://github.com/AliceO2Group/O2Physics/blob/3301cc27ae21395c526e867ce7ef6ee98ed41d84/PWGHF/D2H/Tasks/taskDs.cxx#L84 as example.